Staking pallet

akildemir opened the pull request:
- Implements stake/unstake
 - Allocate/deallocate to specific network validator set
 - Validator set rotation for serai network

Commit:
initial staking pallet

Commit:
add staking pallet to runtime

Commit:
support session rotation for serai

Commit:
optimizations & cleaning

Commit:
Merge branch 'develop' of https://github.com/serai-dex/serai into staking-pallet

Commit:
fix deny

Commit:
add serai network to initial networks

Commit:
a few tweaks & comments

kayabaNerve commented on substrate/client/tests/validator_sets.rs:
```diff
@@ -38,7 +38,7 @@ serai_test!(
         .await
         .unwrap(),
-      [NetworkId::Bitcoin, NetworkId::Ethereum, NetworkId::Monero]
+      [NetworkId::Bitcoin, NetworkId::Ethereum, NetworkId::Monero, NetworkId::Serai]
```
Serai should be first, not last in these cases (a few follow in the files changed view).

kayabaNerve commented on substrate/staking/pallet/LICENSE:
```diff
@@ -0,0 +1,15 @@
+AGPL-3.0-only license
+
+Copyright (c) 2022 Luke Parker
```
It's now 2023, so this would need an extension.

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+  #[pallet::error]
+  pub enum Error<T> {
+    BondUnavailable,
+    InSufficientAllocation,
```
"Insufficient"

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+      Self::allocate_internal(&account, amount)?;
+
+      // add to participants list for the network
+      let result = VsPallet::<T>::add_participant(account, Amount(amount), network);
```
Hm. It's not adding if they were already added.

`increase_allocation`? What do you think would be a more apt name?

akildemir replied:
`handle_allocation`? since then inside of the function, we either add or increase.

akildemir replied:
Actually no handle is very general, it might also mean the decrease. So I guess best is `increase_allocation`, since also can assume that in the case of add, validator had `0` allocation an now has more than `0`.

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+      // add to participants list for the network
+      let result = VsPallet::<T>::add_participant(account, Amount(amount), network);
+      if result.is_err() {
+        Self::deallocate_internal(&account, amount).unwrap();
```
If this returns an error, all state changes from the call will be discarded (since https://github.com/paritytech/substrate/pull/11431). Accordingly, you should simply be able to `?` the error, without this restoration code (which may have its own bug).

akildemir replied:
you mean ` VsPallet::<T>::add_participant(account, Amount(amount), network)` (directly returns the `DispatchResult` like in the `deallocate`) and this will revert the allocation we did if it returns error?

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+    /// Allocate `amount` to a given validator set.
+    #[pallet::call_index(3)]
+    #[pallet::weight((0, DispatchClass::Operational))] // TODO
+    pub fn deallocate(
```
Incorrect doc comment.

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+      // we can't directly deallocate here, since the leaving validator
+      // will be removed after the next session. We only deallocate then
+      // on `end_session` for the right index.
+      VsPallet::<T>::maybe_remove_participant(account, Amount(amount), network)
```
`decrease_allocation`?

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+
+  /// Call order is end_session(i - 1) -> start_session(i) -> new_session(i + 1)
+  /// new_session(i + 1) is called immediately after start_session(i) returns then
+  /// we wait until the session ends then get a call to end_session(i) and so on.
```
The `///` signifies doc comment, yet this won't generate documentation. `//`?

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+      // Returning None will keep the previous set going.
+      if next_validators.is_empty() {
+        return None;
+      }
```
Why will next_validator_set return an empty list?

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+      for (account, amount, _) in deallocating_validators {
+        // we can unwrap because we are not deallocating more than allocated.
+        <Self as AllocatedStaking<T>>::deallocate(&account, amount.0).unwrap();
+      }
```
This is O(n) (meaning `end_session` can be nuked into an extremely expensive call).

Can we have it where, once a session expires (Serai or external network), the validator submits a TX to refresh their allocation (instead of handling it here for them)? We can even do it automatically when `allocate`/`unstake` is called so multiple TXs aren't needed.

It does require storing deallocations as:

`Queued (amount, session)`

And if further deallocation is queued in a new session, shifting the existing deallocation from `Queued (amount, session)` to `Complete amount`.

akildemir replied:
Hmm yes that is possible. We can do the the refresh whenever we get a call to `allocate`/`unstake`.  We wouldn't need to store additional info for it(as far as immediately thinking about it.) we could just check the active bond amount, and release any amount that is more than necessary. The main concern would be future complications due to not having refreshed state immediately.

kayabaNerve commented on substrate/staking/pallet/src/lib.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  impl<T: Config> AllocatedStaking<T> for Pallet<T> {
```
Why did we introduce this new trait? Because else we had validator-sets pallet import staking pallet and staking pallet implement validator-sets pallet?

akildemir replied:
I was actually wondering that myself lol. Why do we have it? it was just there in the original staking folder so I just imported it lol. But will delete because doesn't seem necessary now 👍

akildemir replied:
uhh but that makes the whole `staking/primitives` crate empty? Should we nuke the whole crate then?

kayabaNerve commented on substrate/validator-sets/pallet/Cargo.toml:
```diff
@@ -39,13 +41,17 @@ std = [
   "frame-system/runtime-benchmarks",
   "frame-support/runtime-benchmarks",
+
+	"pallet-babe/runtime-benchmarks",
```
Is this a tab instead of two spaces?

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -63,13 +70,51 @@ pub mod pallet {
+  #[pallet::storage]
+  #[pallet::getter(fn joining_validators)]
+  #[pallet::unbounded]
+  pub type JoiningValidators<T: Config> =
```
What about validators who aren't newly joining yet have allocated more bond?

akildemir replied:
They are added to the same set, yea renaming is necessary. It should be `AllocatingValidators` .

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -63,13 +70,51 @@ pub mod pallet {
+
+  #[pallet::storage]
+  #[pallet::getter(fn current_session)]
+  pub type CurrentSessionIndex<T> = StorageValue<_, Session, ValueQuery>;
```
https://paritytech.github.io/polkadot-sdk/master/pallet_session/pallet/struct.Pallet.html#method.current_index ?

akildemir replied:
I mean we would be importing the whole pallet to vs pallet just for this call? ValidatorSet pallet doesn't deal with session pallet at all currently. But don't mind, whichever is best.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -63,13 +70,51 @@ pub mod pallet {
+    /// Validator wasn't registered or active.
+    NonExistentValidator,
+    /// Trying to deallocate more than allocated.
+    InSufficientAllocation,
```
"Insufficient"

kayabaNerve commented on substrate/runtime/src/lib.rs:
```diff
@@ -176,6 +178,20 @@ impl Contains<RuntimeCall> for CallFilter {
+    }
+
+    if let RuntimeCall::Session(call) = call {
+      return matches!(call, session::Call::set_keys { .. } | session::Call::purge_keys {});
```
Do we bother exposing purge_keys? Seems like something it only makes sense for us to call internally.

akildemir replied:
umm right. Purge keys takes effect in the next session, so exposing it would remove a validator from the set in the session pallet although we think it is  still in the set in the vs pallet. 👍

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -97,17 +142,32 @@ pub mod pallet {
+      // This is only approximation since slots != blocks.
+      let next_start: u64 = Babe::<T>::next_epoch().start_slot.into();
+      if next_start - 100 == n.saturated_into::<u64>() {
+        // TODO: only for serai or for all networks?
```
I'd just do all of them here.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+      // Since this participant will be included after the next session, save it for the set
+      // where it will be active.
+      let participant_set =
+        ValidatorSet { session: Session(Self::current_session().0 + 2), network };
```
`+ 1` if not Serai, as only Serai has this limitation?

Though also, it's frustrating that it'll take two weeks to join the Serai validator set. Do we want to consider halving the discussed session time from 7 days to 3.5, yet only rotating external networks every other session? That'd set a maximum of 7 days to join the Serai set (not 2 weeks), while maintaining the 7 day interval for external networks.

akildemir replied:
yea we can set the rotation time to 3.5 days, but it might still be better for other networks to follow the serai net? that would purify the design, since all the commands would be coming from the session pallet. Otherwise  we are kinda looking at  implementing our own session pallet for other networks. Which if we were gonna do that, we can also use it for the serai net as well. But yea migration taking 6 hrs every 3.5 days is also a problem in that case.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+        ValidatorSet { session: Session(Self::current_session().0 + 2), network };
+      JoiningValidators::<T>::try_mutate_exists(participant_set, |existing| {
+        if existing.is_some() {
+          existing.as_mut().unwrap().push((account, amount));
```
They may already be present?

kayabaNerve replied:
Also, this is an O(n) algorithm. If we move to keying off ValidatorSet *and account*, I believe you can have O(1) insertion (well, the complexity would be the underlying DBs which should be O(log n), yet we can handwave it) yet still trivially do an O(n) iteration over the storage prefix. We should at least `TODO` for this.

akildemir replied:
I didn't understand why it would be O(n), we are not iterating through anything if I am not mistaken.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+      }
+
+      // if the remaining bond is not enough remove from the validator set and
+      // deallocate all bond instead of just "amount" since they will be leaving.
```
I'd error, unless they deallocate the entire amount. Then someone trying to deallocate 10 doesn't undo a 100k bond position.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+        None => {
+          // check whether it is still in the joining set before getting active
+          let joining_set = JoiningValidators::<T>::get(key);
+          *joining_set.iter().find(|p| p.0 == account).ok_or(Error::<T>::NonExistentValidator)?
```
This seems error-prone (removing EITHER from current or joining, yet still going through `DeallocatingValidators`). If they have yet to join, can't we deallocate in place without DV?

akildemir replied:
uhh we should be able to through a callback? The problem was since staking -> vs-pallet, we couldn't call the staking pallet function. Idk how desirable a callback tho.

kayabaNerve commented on substrate/validator-sets/pallet/src/lib.rs:
```diff
@@ -127,6 +187,227 @@ pub mod pallet {
+      // append them all together
+      joining.extend(prev.participants);
+      let last_add_index = joining.len() - 1;
+      joining.extend(deallocating);
```
Why is `deallocating` merged into `joining`?

akildemir replied:
all 3 vectors merged into 1 to not repeat [following](https://github.com/akildemir/serai/blob/baaf96f59c06b1157e75059d275cc69de5625f1a/substrate/validator-sets/pallet/src/lib.rs#L348) block of code.

Commit:
fix some pr comments

Commit:
rebase to develop latest

kayabaNerve committed:
Rewrite validator-sets with logarithmic algorithms

Uses the fact the underlying DB is sorted to achieve sorting of potential
validators by stake.

Removes release of deallocated stake for now.
